### DISCUSSION TITLE
questions about Mocap-Drones bad accuracy

### DISCUSSION THREAD

**Discussion by wykisthebest** [jyjblrd/Mocap-Drones#23] [Q&A] [answered]
I built a motion capture environment based on the code on github.Used 4 ps3 eye cameras and also carried out IR modification，as shown below。
![image](https://github.com/jyjblrd/Mocap-Drones/assets/95999160/70b40803-730c-40d9-b0f0-be6f58745ea4)
I used an infrared lamp bead, as shown below
![image](https://github.com/jyjblrd/Mocap-Drones/assets/95999160/9d132141-f0f8-4186-b754-fce95c9b9f9b)
Front-end web pages can also be displayed.
![image](https://github.com/jyjblrd/Mocap-Drones/assets/95999160/21676442-d31c-4794-9691-14afddda2bf0)
In the Collect points for camera pose calibration step, I used the infrared lamp beads above to move in the space, capture about 5,000 points, and then perform calculations.
![image](https://github.com/jyjblrd/Mocap-Drones/assets/95999160/baefe6cc-a581-4dde-ac4f-205aad259281)
But the final calculation result is not very good.
![image](https://github.com/jyjblrd/Mocap-Drones/assets/95999160/018a445a-2f50-4192-b894-7fa7b6007523)
I wonder if I missed any steps that caused this to happen?

**Comment by jyjblrd** [comment, marked as answer]
Hi, 

The camera calibration can be a bit finicky sometimes, and you might need to do it a few times before getting a good calibration. A few tips are:
1. Make sure that your camera intrinsics in the `camera_param.json` file are correct
2. When collecting point for calibration, make sure that they you are moving the marker around in 3d space throughout the whole field of view of all the cameras
3. I've noticed that I got a better calibration when using less points, not sure why? The best results were when I only blinked my IR led in a few places around the room, and ran the calibration on these 20-30 points. Maybe the optimization algorithm isn't very good at rejecting outliers? That might be something to look into.
4. Make sure you don't accidentally move your camera after calibration! That might be what happened to make your 4th camera's epipolar line not line up with the point.
5. Play around with the [optimizer's parameters](https://github.com/jyjblrd/Mocap-Drones/blob/8d5169e466d6ece850094e5713f7ec2fbbe4cff0/computer_code/api/helpers.py#L287-L289) to try find a loss function, ftol, or xtol that works better.

Hope that helps!

**Comment by wykisthebest** [comment]
Thank you so much for patiently answering my questions! 

I will study the information you provided and attempt to implement it in my project. If I encounter any new challenges during the process, I will reach out to you again for guidance.

Once again, thank you for your support and guidance!

**Reply by muzilyd** [reply]
I am also reproducing this project. Do you have WeChat? I would like to inquire about some questions

**Reply by jyjblrd** [reply]
@muzilyd Please feel free to ask the questions here, and I can help answer them. It also means that others can also benefit from the answers

**Reply by muzilyd** [reply]
@jyjblrd Ok, thank you very much. I have successfully configured PS3eye and installed all required dependencies as required, as shown in the following figure
<img width="391" alt="0a0517de38159dda84ca4cb0861369ec" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/1abbe7af-315a-4368-a31c-3185162b7004">
However, when I tried to run index.py after running yarn run dev, I encountered the following error. What should I do to resolve it.
<img width="665" alt="{F935F648-5937-499c-B59C-E4114D4913C9}" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/bbf25d3c-e00b-4295-b1d2-620396e036ab">

**Reply by jyjblrd** [reply]
Thats a strange error, are you sure that you havent edited anything in the code? maybe this is the issue: [https://stackoverflow.com/a/339220/9776627](https://stackoverflow.com/a/339220/9776627)

**Reply by muzilyd** [reply]
@jyjblrd I added __init__.py to the folder, but it didn't work
<img width="171" alt="{395FEC8F-E298-4852-87EB-77F33DA7234F}" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/c69fe213-6a12-49d9-b708-215a776cf511">
I annotated all sections in "index. py" that use serial, because I was unable to successfully determine which port the camera is using on a Windows computer, but I feel that this is not the reason for "no model named index"
<img width="366" alt="{0F106EC9-056C-4111-B147-33444020E24F}" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/1bab913d-e0b6-4a05-a54d-3357c184f695">
<img width="550" alt="{086012CF-6AC3-4ebf-9501-B40148BC49AF}" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/64c7b901-037d-496d-b3c6-f050b6a489ea">

**Reply by jyjblrd** [reply]
Try to clone the repo so there are no changes, and then edit the serial line to `ser = None`. Then run `python3 api/index.py`

**Reply by muzilyd** [reply]
@jyjblrd After I re cloned the repo, I found that test.py was no longer working. 
<img width="859" alt="{A14A210C-3117-4384-B64A-BDD496B2833B}" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/3603a5d7-81d8-49bd-98a3-7cd4009bf38e">
However, I may have found a problem in the previous clone project. When I replaced "test.py" with "test1. py", I found that it had the same problem. However, I don't know how to modify the file "index. py" because it is originally called index.
<img width="143" alt="{641FA0F4-49EE-4d0d-AF35-1C8E5AC0C2A2}" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/7605f468-8c3d-4137-8b46-f3f4e557f79d">
<img width="254" alt="{13C641D0-61AD-469d-AFFC-936C10CCBF24}" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/77f7cad1-ffd9-4d33-8760-1b6471469e1a">
<img width="195" alt="{4C04585F-040F-4254-ACCE-D2C36DFC10E3}" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/e4ce1a1e-1070-4ef4-b6d5-d84b0b5cd721">

**Reply by muzilyd** [reply]
@jyjblrd Thank you very much, I have solved this question, Although I don't know why, when I use "start debugging", the issue of "no model named index" appears. When I use "run without debugging", it succeeds and is displayed on the UI interface, as shown in the following figure. Thank you again for your open-source project. I have learned a lot from it. Then, I will continue to complete the remaining part.
<img width="389" alt="2ef6b461140856147fe260c456def6e7" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/dd6b9c9e-d002-42cf-9232-fc7d323e6b47">
<img width="665" alt="85a5cf7e00a49de40a953eb41d2a0924" src="https://github.com/jyjblrd/Mocap-Drones/assets/73511912/efa7b021-e0e4-4350-8476-c22b23bb74f0">
